### PR TITLE
Use Duration instead of int in ConfigurationProperties

### PR DESCRIPTION
**Pull request by jandusil**

### CODE REVIEW

**Comment by jandusil** [comment]
I am not sure about whether to create new migration file, or include it to PowerAuth-Server-1.5.0.md

File: docs/Configuration-Properties.md [hunk 1]
```diff
@@ -28,9 +28,9 @@ The PowerAuth Server uses the following public configuration properties:
 
 | Property                                                      | Default   | Note                                                                                    |
 |---------------------------------------------------------------|-----------|-----------------------------------------------------------------------------------------|
-| `powerauth.service.crypto.activationValidityInMilliseconds`   | `120000`  | Default activation validity period in miliseconds                                       |
+| `powerauth.service.crypto.activationValidityTime`   | `PT2M`  | Default activation validity period (ISO 8601 Duration format) |
```
Comment (banterCZ) [inline docs/Configuration-Properties.md:31]: May we use the variant without `PT` prefix? Spring Boot can do that, see https://docs.spring.io/spring-boot/docs/2.1.13.RELEASE/reference/html/boot-features-external-config.html#boot-features-external-config-conversion-duration

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java [hunk 2]
```diff
@@ -196,21 +206,26 @@ public class PowerAuthServiceConfiguration {
...
     /**
-     * Token timestamp validity to future in milliseconds, checked before validating the token.
+     * Specifies the validity duration to future of token timestamps, checked before token validation.
+     * <p>
+     * The value is specified in the ISO-8601 duration format PTnHnMn.nS.
      */
     @Value("${powerauth.service.token.timestamp.forward.validity}")
-    @Min(0)
-    private long tokenTimestampForwardValidityInMilliseconds;
+    @DurationMin(millis = 0)
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java:226]: What is the difference between this and the previous format `@Min(0)`?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ActivationServiceBehavior.java [hunk 3]
```diff
@@ -686,7 +687,7 @@ public InitActivationResponse initActivation(String applicationId, String userId
             // Get activation expiration date from request or from constants, if not provided
             Date timestampExpiration = activationExpireTimestamp;
             if (timestampExpiration == null) {
-                timestampExpiration = new Date(timestamp.getTime() + powerAuthServiceConfiguration.getActivationValidityBeforeActive());
+                timestampExpiration = Date.from(Instant.now().plus(powerAuthServiceConfiguration.getActivationValidityBeforeActive()));
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ActivationServiceBehavior.java:690]: Are you sure that `timestamp` (before) is the same as `Instant.now()`? I think that there is a little time shift.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/CallbackUrlBehavior.java [hunk 4]
```diff
@@ -431,7 +431,7 @@ private synchronized RestClient getRestClient(CallbackUrlEntity callbackUrlEntit
     private RestClient initializeRestClient(CallbackUrlEntity callbackUrlEntity) throws RestClientException {
         final DefaultRestClient.Builder builder = DefaultRestClient.builder();
         if (configuration.getHttpConnectionTimeout() != null) {
-            builder.connectionTimeout(configuration.getHttpConnectionTimeout());
+            builder.connectionTimeout((int) configuration.getHttpConnectionTimeout().toMillis());
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/CallbackUrlBehavior.java:434]: Not much dangerous here, but better to use `Math#toIntExact()`.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java (hunk 2 above)
Comment (jandusil) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java:226]: The @Min apparently wont work on type Duration

File: docs/Configuration-Properties.md (hunk 1 above)
Comment (jandusil) [reply docs/Configuration-Properties.md:31]: While it is true that we can omit the 'PT' prefix and still have Spring Boot interpret the duration correctly, I believe we should consider the implications for broader public documentation and usage. By adhering to the ISO 8601 Duration format in its complete form, including the 'PT' prefix, we ensure that our format is standard-compliant and immediately recognizable to anyone familiar with ISO 8601.

Comment (banterCZ) [reply docs/Configuration-Properties.md:31]: Standards are fine. Let's wait for another opinion and we will decide.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/CallbackUrlBehavior.java [hunk 5]
```diff
@@ -431,7 +431,7 @@ private synchronized RestClient getRestClient(CallbackUrlEntity callbackUrlEntit
     private RestClient initializeRestClient(CallbackUrlEntity callbackUrlEntity) throws RestClientException {
         final DefaultRestClient.Builder builder = DefaultRestClient.builder();
         if (configuration.getHttpConnectionTimeout() != null) {
-            builder.connectionTimeout(configuration.getHttpConnectionTimeout());
+            builder.connectionTimeout(Math.toIntExact(configuration.getHttpConnectionTimeout().toMillis()));
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/CallbackUrlBehavior.java:434]: 🤔 I have opened another issue, what we may improve in the next step. https://github.com/wultra/lime-java-core/issues/201

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/replay/ReplayPersistenceService.java [hunk 6]
```diff
@@ -70,7 +70,7 @@ public boolean uniqueValueExists(final String uniqueValue) {
      * @return Whether unique value was added successfully.
      */
     public boolean persistUniqueValue(final UniqueValueType type, final String uniqueValue) {
-        final Instant expiration = Instant.now().plus(config.getRequestExpirationInMilliseconds(), ChronoUnit.MILLIS);
+        final Instant expiration = Instant.now().plus(config.getRequestExpiration().toMillis(), ChronoUnit.MILLIS);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/replay/ReplayPersistenceService.java:73]: Doesn't it work in a simpler form?
```suggestion
        final Instant expiration = Instant.now().plus(config.getRequestExpiration());
```

**Review by banterCZ — APPROVED** [review]
Good job (except my personal opinion about the duration format).
Let's wait for the fix of crypto 3.1 and run backend tests before merging.

**Pull request merged**